Assets/LeapMotionModules/Interaction/InteractionBehaviour.cs

```diff
@@ -0,0 +1,711 @@
+          }
+          else {
+            OnMultiGraspHold(_graspingHandsBuffer);
+          }
+        }
+      }
+      if (_graspCount == 0) {
+        if (_graspCountLastFrame > 0) {
+          OnObjectGraspEnd();
+        }
+      }
+      if (_graspCount <= 1) {
+        if (_graspCountLastFrame > 1) {
+          OnMultiGraspEnd(_lastFrameGraspingHands);
+        }
+      }
+
+      _graspCountLastFrame = _graspCount;
+      // Keep track of grasping hands from last frame; these must be independent
+      // Hand objects due to pooling by the service provider, but care must be taken
+      // to not allocate garbage every frame!
+      while (_lastFrameGraspingHands.Count < _graspingHandsBuffer.Count) {
+        if (_lastFrameGraspingHandObjectStorage.Count > 0) {
+          _lastFrameGraspingHands.Add(_lastFrameGraspingHandObjectStorage.Pop());
+        }
+        else {
+          _lastFrameGraspingHandObjectStorage.Push(new Hand());
+        }
+      }
+      while (_lastFrameGraspingHands.Count > _graspingHandsBuffer.Count) {
+        _lastFrameGraspingHandObjectStorage.Push(_lastFrameGraspingHands[_lastFrameGraspingHands.Count - 1]);
+        _lastFrameGraspingHands.RemoveAt(_lastFrameGraspingHands.Count - 1);
+      }
+      for (int i = 0; i < _graspingHandsBuffer.Count; i++) {
+        _lastFrameGraspingHands[i].CopyFrom(_graspingHandsBuffer[i]);
+      }
+    }
+
+    public override void GraspBegin(Hand hand) {
+      if (isGrasped && !allowMultiGrasp) {
+        interactionManager.ReleaseObjectFromGrasp(this);
+      }
+
+      InteractionHand graspingIntHand = interactionManager.GetInteractionHand(hand);
+      _graspingIntHands.Add(graspingIntHand);
+      _graspCount++;
+
+      // SnapToHand(hand); // TODO: When you grasp an object, snap the object into a good holding position.
+
+      if (moveObjectWhenGrasped) {
+        GraspedPoseController.AddHand(graspingIntHand);
+      }
+
+      // Set kinematic state based on grasping hold movement type
+      if (_graspCount == 1) {
+        _wasKinematicBeforeGrab = rigidbody.isKinematic;
+        switch (graspedMovementType) {
+          case GraspedMovementType.Inherit: break; // no change
+          case GraspedMovementType.Kinematic:
+            rigidbody.isKinematic = true; break;
+          case GraspedMovementType.Nonkinematic:
+            rigidbody.isKinematic = false; break;
+        }
+      }
+
+      OnGraspBegin(hand);
+    }
+
+    public override void GraspHold(Hand hand) {
+      if (moveObjectWhenGrasped) {
+        Vector3 origPosition = rigidbody.position; Quaternion origRotation = rigidbody.rotation;
+        Vector3 newPosition; Quaternion newRotation;
+        GraspedPoseController.GetGraspedPosition(out newPosition, out newRotation);
+
+        IGraspedMovementController holdingMovementController = rigidbody.isKinematic ?
+                                                                 (IGraspedMovementController)_kinematicHoldingMovement
```

Comment:
Wow, TIL that you can't use a ternary expression in this situation without casting.  The reason I thought you could remove it was that Visual Studio was annotating that both casts could be removed, but as soon as you remove one of them it stops saying you can remove the other one XD